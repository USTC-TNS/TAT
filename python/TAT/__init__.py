import cppyy
cppyy.load_library("blas")
cppyy.load_library("lapack")
cppyy.include("../include/TAT/TAT.hpp")
std = cppyy.gbl.std
TAT = cppyy.gbl.TAT
