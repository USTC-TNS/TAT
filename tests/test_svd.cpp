#include <TAT/TAT.hpp>
#include <gtest/gtest.h>

#define t_edge(...) \
    { {__VA_ARGS__}, true }
#define f_edge(...) \
    { {__VA_ARGS__}, false }

template<typename T>
auto check_unitary(const T& tensor, const std::string& name, const std::string& name_prime) {
    auto pairs = std::unordered_set<std::pair<std::string, std::string>>();
    for (auto n : tensor.names()) {
        if (n != name) {
            pairs.insert({n, n});
        }
    }
    auto conjugated = tensor.conjugate(true).edge_rename({{name, name_prime}});
    auto product = tensor.contract(conjugated, pairs);
    auto identity = product.same_shape().identity_({{name, name_prime}});
    if constexpr (T::symmetry_t::is_fermi_symmetry) {
        // some sign maybe wrong
        product.transform_([](auto x) { return std::abs(x); });
        identity.transform_([](auto x) { return std::abs(x); });
    }
    return (product - identity).template norm<-1>();
}

TEST(test_svd, no_symmetry) {
    auto a = TAT::Tensor<double, TAT::NoSymmetry>{{"A", "B", "C", "D"}, {2, 3, 4, 5}}.range_();
    auto [u, s, v] = a.svd({"C", "A"}, "E", "F", "U", "V");
    ASSERT_NEAR(check_unitary(u, "E", "E'"), 0, 1e-8);
    ASSERT_NEAR(check_unitary(v, "F", "F'"), 0, 1e-8);
    auto b = v.contract(s, {{"F", "V"}}).contract(u, {{"U", "E"}});
    ASSERT_NEAR((a - b).norm<-1>(), 0, 1e-8);
}

TEST(test_svd, no_symmetry_cut) {
    auto a = TAT::Tensor<double, TAT::NoSymmetry>{{"A", "B", "C", "D"}, {5, 4, 3, 2}}.range_();
    auto [u, s, v] = a.svd({"B", "D"}, "E", "F", "U", "V", 2ul);
    ASSERT_NEAR(check_unitary(u, "E", "E'"), 0, 1e-8);
    ASSERT_NEAR(check_unitary(v, "F", "F'"), 0, 1e-8);
    auto b = v.contract(s, {{"F", "V"}}).contract(u, {{"U", "E"}});
    ASSERT_NEAR((a - b).norm<-1>(), 0, 1e-8);
}

TEST(test_svd, u1_symmetry) {
    auto a = (TAT::Tensor<double, TAT::U1Symmetry>{
        {"A", "B", "C", "D"},
        {t_edge({-1, 1}, {0, 1}, {-2, 1}), f_edge({0, 1}, {1, 2}), f_edge({0, 2}, {1, 2}), t_edge({-2, 2}, {-1, 1}, {0, 2})}}
                  .range_());
    auto [u, s, v] = a.svd({"B", "D"}, "E", "F", "U", "V");
    ASSERT_NEAR(check_unitary(u, "E", "E'"), 0, 1e-8);
    ASSERT_NEAR(check_unitary(v, "F", "F'"), 0, 1e-8);
    auto b = v.contract(s, {{"F", "V"}}).contract(u, {{"U", "E"}});
    ASSERT_NEAR((a - b).norm<-1>(), 0, 1e-8);
}

TEST(test_svd, u1_symmetry_cut) {
    auto a = (TAT::Tensor<double, TAT::U1Symmetry>{
        {"A", "B", "C", "D"},
        {t_edge({-1, 1}, {0, 1}, {-2, 1}), f_edge({0, 1}, {1, 2}), f_edge({0, 2}, {1, 2}), t_edge({-2, 2}, {-1, 1}, {0, 2})}}
                  .range_());
    auto [u, s, v] = a.svd({"C", "A"}, "E", "F", "U", "V", 7ul);
    ASSERT_NEAR(check_unitary(u, "E", "E'"), 0, 1e-8);
    ASSERT_NEAR(check_unitary(v, "F", "F'"), 0, 1e-8);
    auto b = v.contract(s, {{"F", "V"}}).contract(u, {{"U", "E"}});
    ASSERT_NEAR((a - b).norm<-1>(), 0, 1e-8);
}

TEST(test_svd, fermi_symmetry) {
    auto a = (TAT::Tensor<double, TAT::FermiSymmetry>{
        {"A", "B", "C", "D"},
        {t_edge({-1, 1}, {0, 1}, {-2, 1}), f_edge({0, 1}, {1, 2}), f_edge({0, 2}, {1, 2}), t_edge({-2, 2}, {-1, 1}, {0, 2})}}
                  .range_());
    auto [u, s, v] = a.svd({"C", "A"}, "E", "F", "U", "V");
    ASSERT_NEAR(check_unitary(u, "E", "E'"), 0, 1e-8);
    ASSERT_NEAR(check_unitary(v, "F", "F'"), 0, 1e-8);
    auto b = v.contract(s, {{"F", "V"}}).contract(u, {{"U", "E"}});
    ASSERT_NEAR((a - b).norm<-1>(), 0, 1e-8);
}

TEST(test_svd, fermi_symmetry_cut) {
    auto a = (TAT::Tensor<double, TAT::FermiSymmetry>{
        {"A", "B", "C", "D"},
        {t_edge({-1, 1}, {0, 1}, {-2, 1}), f_edge({0, 1}, {1, 2}), f_edge({0, 2}, {1, 2}), t_edge({-2, 2}, {-1, 1}, {0, 2})}}
                  .range_());
    auto [u, s, v] = a.svd({"B", "D"}, "E", "F", "U", "V", 8ul);
    ASSERT_NEAR(check_unitary(u, "E", "E'"), 0, 1e-8);
    ASSERT_NEAR(check_unitary(v, "F", "F'"), 0, 1e-8);
    auto b = v.contract(s, {{"F", "V"}}).contract(u, {{"U", "E"}});
    ASSERT_NEAR((a - b).norm<-1>(), 0, 1e-8);
}

TEST(test_svd, no_symmetry_cut_too_small) {
    auto a = TAT::Tensor<double, TAT::NoSymmetry>{{"A", "B"}, {2, 2}}.zero_();
    a.at(std::vector<TAT::Size>{0, 0}) = 1;
    auto [u, s, v] = a.svd({"B"}, "E", "F", "U", "V", 8ul);
    ASSERT_NEAR(check_unitary(u, "E", "E'"), 0, 1e-8);
    ASSERT_NEAR(check_unitary(v, "F", "F'"), 0, 1e-8);
    auto b = v.contract(s, {{"F", "V"}}).contract(u, {{"U", "E"}});
    ASSERT_NEAR((a - b).norm<-1>(), 0, 1e-8);
    ASSERT_EQ(s.storage().size(), 1);
}

TEST(test_svd, fermi_symmetry_cut_too_small) {
    auto a = TAT::Tensor<double, TAT::FermiSymmetry>{{"A", "B"}, {{{0, 1}, {+1, 1}}, {{-1, 1}, {0, 1}}}}.range_(0, 1);
    auto [u, s, v] = a.svd({"B"}, "E", "F", "U", "V", 8ul);
    ASSERT_NEAR(check_unitary(u, "E", "E'"), 0, 1e-8);
    ASSERT_NEAR(check_unitary(v, "F", "F'"), 0, 1e-8);
    auto b = v.contract(s, {{"F", "V"}}).contract(u, {{"U", "E"}});
    ASSERT_NEAR((a - b).norm<-1>(), 0, 1e-8);
    ASSERT_EQ(s.storage().size(), 1);
    ASSERT_EQ(s.edges(0).segments().size(), 1);
}
