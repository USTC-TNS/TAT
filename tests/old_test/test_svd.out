{[32mnames[0m:[A,B,C,D],[32medges[0m:[2,3,4,5],[32mblocks[0m:[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119]}
{[32mnames[0m:[A,C,E],[32medges[0m:[2,4,8],[32mblocks[0m:[-0.129192,-0.545039,0.00264755,-0.0889997,0.0436829,-0.302797,-0.266546,-0.71671,-0.154407,-0.492887,-0.0594364,-0.0108678,-0.0736774,-0.0715212,-0.55916,0.637469,-0.179622,-0.440734,-0.138556,0.0473052,0.0972223,-0.349421,0.75182,0.235105,-0.204836,-0.388581,0.154105,-0.0937559,0.0321574,0.857496,0.177488,-0.0816842,-0.43177,0.0807945,0.163343,0.874423,0.0833733,0.00501036,-0.0695334,-0.0627691,-0.456985,0.132947,-0.229087,-0.115946,-0.835522,-0.0089162,0.0612196,-0.0751018,-0.4822,0.1851,0.712518,-0.394129,0.147388,-0.201433,0.0090828,0.088437,-0.507414,0.237253,-0.605533,-0.218029,0.505376,0.0715812,-0.10437,-0.0247462]}
{[32mnames[0m:[F,B,D],[32medges[0m:[8,3,5],[32mblocks[0m:[-0.177428,-0.180824,-0.184219,-0.187615,-0.19101,-0.245338,-0.248734,-0.252129,-0.255525,-0.25892,-0.313248,-0.316644,-0.320039,-0.323435,-0.32683,0.394062,0.37868,0.363298,0.347916,0.332534,0.0864199,0.0710378,0.0556557,0.0402736,0.0248915,-0.221222,-0.236605,-0.251987,-0.267369,-0.282751,-0.852095,0.254727,0.0973441,0.218279,0.15109,0.00383678,0.135653,0.0550223,0.0225419,0.16445,-0.251663,0.0638962,0.0518081,-0.0162014,-0.0986887,-0.0117278,-0.596974,0.345133,0.262481,0.240903,-0.0998087,-0.0230353,0.0675831,-0.169087,-0.134826,-0.277332,-0.000224406,-0.184604,0.457396,0.124123,0.198902,-0.302611,0.00148003,-0.100566,0.071835,-0.221725,0.158088,-0.181897,0.0675722,0.491704,-0.384414,0.278373,0.400025,-0.215249,-0.261516,-0.147772,-0.268385,0.281625,-0.0785091,0.0790041,0.129658,0.118874,-0.20931,0.142709,0.0848212,0.398051,-0.633462,0.31811,0.0105882,-0.226002,-0.113896,-0.388612,0.0562622,0.270025,0.0743219,0.237671,0.0569561,0.0498124,-0.0188509,-0.217975,0.36669,0.418133,-0.103639,-0.553654,-0.133245,0.0955308,0.0407738,-0.481474,0.473849,-0.0741949,0.117067,0.127383,-0.111045,-0.225665,-0.0904979,0.0761918,0.0164473,0.164979,0.373689,-0.503033]}
{[32mnames[0m:[U,V],[32medges[0m:[8,8],[32mblocks[0m:[749.943,0,0,0,0,0,0,0,0,80.0374,0,0,0,0,0,0,0,0,6.66284e-14,0,0,0,0,0,0,0,0,2.37141e-14,0,0,0,0,0,0,0,0,1.76226e-14,0,0,0,0,0,0,0,0,6.94321e-15,0,0,0,0,0,0,0,0,5.83549e-15,0,0,0,0,0,0,0,0,1.54614e-15]}
{[32mnames[0m:[A,B,C,D],[32medges[0m:[2,3,4,5],[32mblocks[0m:[8.28257e-16,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119]}
{[32mnames[0m:[A,B,C,D],[32medges[0m:[2,3,4,5],[32mblocks[0m:[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119]}
{[32mnames[0m:[E,A,D],[32medges[0m:[10,2,5],[32mblocks[0m:[-0.141922,-0.146421,-0.15092,-0.15542,-0.159919,-0.411884,-0.416383,-0.420882,-0.425382,-0.429881,0.439754,0.430235,0.420715,0.411196,0.401676,-0.131414,-0.140934,-0.150453,-0.159973,-0.169492,0.103897,0.172575,0.00222105,-0.0213988,-0.314405,0.309457,0.67091,-0.169096,-0.402117,-0.352043,0.463723,-0.392835,0.0300056,0.116038,-0.20805,-0.520504,0.26859,0.0621269,-0.233355,0.414261,0.393269,-0.467082,0.00399786,0.159524,-0.104249,0.247947,-0.0444164,-0.42173,0.517899,-0.285161,0.45152,0.0130526,-0.37933,-0.111494,0.000338761,0.512998,-0.402502,0.0623643,-0.389524,0.242577,-0.177434,0.260301,-0.337224,0.167081,0.11111,-0.0329729,0.15359,-0.696433,0.0701489,0.481833,-0.116538,-0.346916,0.552367,-0.439849,0.360223,0.225361,0.0610639,-0.283535,-0.232746,0.220569,0.0543584,-0.235918,-0.483708,-0.0397866,0.719497,-0.11126,0.29802,0.12104,-0.0601505,-0.262092,0.392544,0.383741,-0.0510564,-0.72938,-0.0172276,-0.239782,0.0662337,-0.0772133,0.30765,-0.0355093]}
{[32mnames[0m:[B,C,F],[32medges[0m:[3,4,10],[32mblocks[0m:[-0.176186,-0.513644,0.802766,0.181444,0.0092894,0.00490106,0.115336,0.00273686,-0.087648,-0.0247392,-0.195255,-0.432223,-0.370066,0.207217,0.541581,0.468893,-0.0461404,-0.103735,-0.0427216,0.230201,-0.214325,-0.350803,-0.260632,-0.371871,-0.358478,-0.0939767,-0.0510079,-0.359514,-0.491061,-0.177766,-0.233395,-0.269382,-0.151483,-0.0351212,-0.0556316,-0.522151,0.0552697,0.0882977,-0.0400943,0.420225,-0.252464,-0.187961,-0.163232,0.15883,-0.101436,0.0451141,-0.0560273,0.0910735,0.274986,-0.771887,-0.271534,-0.10654,-0.133282,-0.122535,0.237693,-0.27103,0.263422,0.176955,0.372131,-0.0926,-0.290604,-0.0251192,-0.0174696,0.0261068,-0.203683,-0.204224,-0.305301,-0.0620953,0.506267,0.241817,-0.309674,0.0563017,-0.0631319,-0.0951225,-0.241978,0.272796,0.104102,0.801937,-0.25387,0.0997604,-0.328743,0.137723,0.0306927,0.229383,-0.478461,0.427412,-0.115551,-0.26056,0.178353,0.226044,-0.347813,0.219143,0.257858,-0.474109,0.373143,0.047372,-0.591761,0.0110409,-0.0737375,-0.0665491,-0.366883,0.300564,0.102177,-0.306812,0.112678,0.143498,0.661554,-0.300432,0.0705146,0.0138671,-0.385952,0.381985,-0.0341983,0.602591,0.165283,-0.318604,-0.033894,-0.0857046,-0.413121,-0.0983729]}
{[32mnames[0m:[U,V],[32medges[0m:[10,10],[32mblocks[0m:[749.622,0,0,0,0,0,0,0,0,0,0,82.983,0,0,0,0,0,0,0,0,0,0,4.92149e-14,0,0,0,0,0,0,0,0,0,0,1.9181e-14,0,0,0,0,0,0,0,0,0,0,1.61879e-14,0,0,0,0,0,0,0,0,0,0,1.06195e-14,0,0,0,0,0,0,0,0,0,0,8.41037e-15,0,0,0,0,0,0,0,0,0,0,4.4572e-15,0,0,0,0,0,0,0,0,0,0,2.94477e-15,0,0,0,0,0,0,0,0,0,0,6.73157e-16]}
{[32mnames[0m:[F,F2],[32medges[0m:[10,10],[32mblocks[0m:[1,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1]}
{[32mnames[0m:[A,B,C,D],[32medges[0m:[2,3,4,5],[32mblocks[0m:[-3.85417e-15,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119]}
{[32mnames[0m:[A,C,E],[32medges[0m:[{-1:1,0:1,-2:1},{0:2,1:2},{1:4,0:4,-1:2}],[32mblocks[0m:{[33m[-1,0,1][0m:[-0.122029,0.719927,0.218629,0.647314,-0.158836,0.664519,-0.187015,-0.705841],[33m[-1,1,0][0m:[-0.252689,0.712297,-0.422826,0.5,-0.295127,0.583405,0.567925,-0.5],[33m[0,0,0][0m:[-0.62999,-0.20385,-0.558173,-0.5,-0.672429,-0.332741,0.432577,0.5],[33m[0,1,-1][0m:[-0.683596,0.729861,-0.729861,-0.683596],[33m[-2,1,1][0m:[-0.674129,-0.111193,-0.692832,0.230585,-0.710936,-0.166601,0.661219,-0.172058]}}
{[32mnames[0m:[U,V],[32medges[0m:[{-1:4,0:4,1:2},{1:4,0:4,-1:2}],[32mblocks[0m:{[33m[-1,1][0m:[108.458,0,0,0,0,4.57202,0,0,0,0,6.76623e-15,0,0,0,0,4.80556e-16],[33m[0,0][0m:[63.4559,0,0,0,0,12.8587,0,0,0,0,0.0310041,0,0,0,0,1.03277e-15],[33m[1,-1][0m:[88.5714,0,0,0.322911]}}
{[32mnames[0m:[F,B,D],[32medges[0m:[{-1:4,0:4,1:2},{0:1,1:2},{-2:2,-1:1,0:2}],[32mblocks[0m:{[33m[-1,1,0][0m:[-0.460596,-0.475956,-0.522037,-0.537397,-0.636788,-0.394739,0.331407,0.573456,0.573378,-0.779118,0.24958,-0.0438404,0.231495,-0.103063,-0.745223,0.616791],[33m[0,0,0][0m:[-0.358892,-0.38805,-0.628687,-0.569653,-0.275588,0.353431,-0.632456,0.632456],[33m[0,1,-1][0m:[-0.570389,-0.628705,0.310613,0.428682,-0.691503,0.566535,0.316228,-0.316228],[33m[1,0,-1][0m:[-0.327408,0.748522],[33m[1,1,-2][0m:[-0.431399,-0.447357,-0.495232,-0.511191,-0.508805,-0.36553,0.0642946,0.20757]}}
{[32mnames[0m:[A,B,C,D],[32medges[0m:[{-1:1,0:1,-2:1},{0:1,1:2},{0:2,1:2},{-2:2,-1:1,0:2}],[32mblocks[0m:{[33m[-1,0,1,0][0m:[0,1,2,3],[33m[-1,1,0,0][0m:[4,5,6,7,8,9,10,11],[33m[-1,1,1,-1][0m:[12,13,14,15],[33m[0,0,0,0][0m:[16,17,18,19],[33m[0,0,1,-1][0m:[20,21],[33m[0,1,0,-1][0m:[22,23,24,25],[33m[0,1,1,-2][0m:[26,27,28,29,30,31,32,33],[33m[-2,1,1,0][0m:[34,35,36,37,38,39,40,41]}}
{[32mnames[0m:[A,B,C,D],[32medges[0m:[{-1:1,0:1,-2:1},{0:1,1:2},{0:2,1:2},{-2:2,-1:1,0:2}],[32mblocks[0m:{[33m[-1,0,1,0][0m:[0,1,2,3],[33m[-1,1,0,0][0m:[4,5,6,7,8,9,10,11],[33m[-1,1,1,-1][0m:[12,13,14,15],[33m[0,0,0,0][0m:[16,17,18,19],[33m[0,0,1,-1][0m:[20,21],[33m[0,1,0,-1][0m:[22,23,24,25],[33m[0,1,1,-2][0m:[26,27,28,29,30,31,32,33],[33m[-2,1,1,0][0m:[34,35,36,37,38,39,40,41]}}
{[32mnames[0m:[A,B,C,D],[32medges[0m:[{-1:1,0:1,-2:1},{0:1,1:2},{0:2,1:2},{-2:2,-1:1,0:2}],[32mblocks[0m:{[33m[-1,0,1,0][0m:[0,1,2,3],[33m[-1,1,0,0][0m:[4,5,6,7,8,9,10,11],[33m[-1,1,1,-1][0m:[12,13,14,15],[33m[0,0,0,0][0m:[16,17,18,19],[33m[0,0,1,-1][0m:[20,21],[33m[0,1,0,-1][0m:[22,23,24,25],[33m[0,1,1,-2][0m:[26,27,28,29,30,31,32,33],[33m[-2,1,1,0][0m:[34,35,36,37,38,39,40,41]}}
{[32mnames[0m:[A,B,C,D],[32medges[0m:[2,3,4,5],[32mblocks[0m:[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119]}
{[32mnames[0m:[A,C,E],[32medges[0m:[2,4,2],[32mblocks[0m:[-0.129192,-0.545039,-0.154407,-0.492887,-0.179622,-0.440734,-0.204836,-0.388581,-0.43177,0.0807945,-0.456985,0.132947,-0.4822,0.1851,-0.507414,0.237253]}
{[32mnames[0m:[F,B,D],[32medges[0m:[2,3,5],[32mblocks[0m:[-0.177428,-0.180824,-0.184219,-0.187615,-0.19101,-0.245338,-0.248734,-0.252129,-0.255525,-0.25892,-0.313248,-0.316644,-0.320039,-0.323435,-0.32683,0.394062,0.37868,0.363298,0.347916,0.332534,0.0864199,0.0710378,0.0556557,0.0402736,0.0248915,-0.221222,-0.236605,-0.251987,-0.267369,-0.282751]}
{[32mnames[0m:[U,V],[32medges[0m:[2,2],[32mblocks[0m:[749.943,0,0,80.0374]}
{[32mnames[0m:[A,B,C,D],[32medges[0m:[2,3,4,5],[32mblocks[0m:[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119]}
{[32mnames[0m:[A,C,E],[32medges[0m:[{-1:1,0:1,-2:1},{0:2,1:2},{1:2,0:3,-1:2}],[32mblocks[0m:{[33m[-1,0,1][0m:[-0.122029,0.719927,-0.158836,0.664519],[33m[-1,1,0][0m:[-0.252689,0.712297,-0.422826,-0.295127,0.583405,0.567925],[33m[0,0,0][0m:[-0.62999,-0.20385,-0.558173,-0.672429,-0.332741,0.432577],[33m[0,1,-1][0m:[-0.683596,0.729861,-0.729861,-0.683596],[33m[-2,1,1][0m:[-0.674129,-0.111193,-0.710936,-0.166601]}}
{[32mnames[0m:[U,V],[32medges[0m:[{-1:2,0:3,1:2},{1:2,0:3,-1:2}],[32mblocks[0m:{[33m[-1,1][0m:[108.458,0,0,4.57202],[33m[0,0][0m:[63.4559,0,0,0,12.8587,0,0,0,0.0310041],[33m[1,-1][0m:[88.5714,0,0,0.322911]}}
{[32mnames[0m:[F,B,D],[32medges[0m:[{-1:2,0:3,1:2},{0:1,1:2},{-2:2,-1:1,0:2}],[32mblocks[0m:{[33m[-1,1,0][0m:[-0.460596,-0.475956,-0.522037,-0.537397,-0.636788,-0.394739,0.331407,0.573456],[33m[0,0,0][0m:[-0.358892,-0.38805,-0.628687,-0.569653,-0.275588,0.353431],[33m[0,1,-1][0m:[-0.570389,-0.628705,0.310613,0.428682,-0.691503,0.566535],[33m[1,0,-1][0m:[-0.327408,0.748522],[33m[1,1,-2][0m:[-0.431399,-0.447357,-0.495232,-0.511191,-0.508805,-0.36553,0.0642946,0.20757]}}
{[32mnames[0m:[A,B,C,D],[32medges[0m:[{-1:1,0:1,-2:1},{0:1,1:2},{0:2,1:2},{-2:2,-1:1,0:2}],[32mblocks[0m:{[33m[-1,0,1,0][0m:[0,1,2,3],[33m[-1,1,0,0][0m:[4,5,6,7,8,9,10,11],[33m[-1,1,1,-1][0m:[12,13,14,15],[33m[0,0,0,0][0m:[16,17,18,19],[33m[0,0,1,-1][0m:[20,21],[33m[0,1,0,-1][0m:[22,23,24,25],[33m[0,1,1,-2][0m:[26,27,28,29,30,31,32,33],[33m[-2,1,1,0][0m:[34,35,36,37,38,39,40,41]}}
{[32mnames[0m:[A,B,C,D],[32medges[0m:[{-1:1,0:1,-2:1},{0:1,1:2},{0:2,1:2},{-2:2,-1:1,0:2}],[32mblocks[0m:{[33m[-1,0,1,0][0m:[0,1,2,3],[33m[-1,1,0,0][0m:[4,5,6,7,8,9,10,11],[33m[-1,1,1,-1][0m:[12,13,14,15],[33m[0,0,0,0][0m:[16,17,18,19],[33m[0,0,1,-1][0m:[20,21],[33m[0,1,0,-1][0m:[22,23,24,25],[33m[0,1,1,-2][0m:[26,27,28,29,30,31,32,33],[33m[-2,1,1,0][0m:[34,35,36,37,38,39,40,41]}}
{[32mnames[0m:[A,B,C,D],[32medges[0m:[{-1:1,0:1,-2:1},{0:1,1:2},{0:2,1:2},{-2:2,-1:1,0:2}],[32mblocks[0m:{[33m[-1,0,1,0][0m:[0,1,2,3],[33m[-1,1,0,0][0m:[4,5,6,7,8,9,10,11],[33m[-1,1,1,-1][0m:[12,13,14,15],[33m[0,0,0,0][0m:[16,17,18,19],[33m[0,0,1,-1][0m:[20,21],[33m[0,1,0,-1][0m:[22,23,24,25],[33m[0,1,1,-2][0m:[26,27,28,29,30,31,32,33],[33m[-2,1,1,0][0m:[34,35,36,37,38,39,40,41]}}
{[32mnames[0m:[E,C,A],[32medges[0m:[{arrow:1,segment:{-1:2,0:4,1:4}},{arrow:0,segment:{0:2,1:2}},{arrow:1,segment:{-1:1,0:1,-2:1}}],[32mblocks[0m:{[33m[-1,1,0][0m:[-0.683596,-0.729861,-0.729861,0.683596],[33m[0,0,0][0m:[-0.640975,-0.683675,-0.194638,-0.290815,0.309526,-0.221391,-0.674878,0.631665],[33m[0,1,-1][0m:[-0.248558,-0.244862,0.567933,0.74498,-0.738096,0.55713,-0.266244,0.273228],[33m[1,0,-1][0m:[-0.122029,-0.158836,-0.719927,-0.664519,-0.681233,0.722706,0.0523102,-0.104317],[33m[1,1,-2][0m:[0.674129,0.710936,-0.111193,-0.166601,-0.0591334,0.100607,-0.727797,0.67579]}}
{[32mnames[0m:[U,V],[32medges[0m:[{arrow:0,segment:{1:2,0:4,-1:4}},{arrow:1,segment:{-1:2,0:4,1:4}}],[32mblocks[0m:{[33m[1,-1][0m:[-88.5714,0,0,-0.322911],[33m[0,0][0m:[62.2698,0,0,0,0,17.7316,0,0,0,0,0.23229,0,0,0,0,0.0623827],[33m[-1,1][0m:[-108.458,0,0,0,0,-4.57202,0,0,0,0,-4.32953e-15,0,0,0,0,-1.1387e-15]}}
{[32mnames[0m:[D,B,F],[32medges[0m:[{arrow:1,segment:{-2:2,-1:1,0:2}},{arrow:0,segment:{0:1,1:2}},{arrow:0,segment:{1:2,0:4,-1:4}}],[32mblocks[0m:{[33m[-2,1,1][0m:[-0.431399,0.508805,-0.495232,-0.0642946,-0.447357,0.36553,-0.511191,-0.20757],[33m[-1,0,1][0m:[0.327408,0.748522],[33m[-1,1,0][0m:[0.577999,-0.311824,-0.444015,-0.609536,0.636392,-0.405155,0.355244,0.551957],[33m[0,0,0][0m:[-0.354458,-0.554877,-0.632261,0.408311,-0.367807,-0.656298,0.526212,-0.396348],[33m[0,1,-1][0m:[0.460596,-0.636788,-0.48039,0.389331,0.522037,0.331407,0.530877,0.579497,0.475956,-0.394739,0.467768,-0.631538,0.537397,0.573456,-0.518255,-0.33729]}}
{[32mnames[0m:[D,C,B,A],[32medges[0m:[{arrow:1,segment:{-2:2,-1:1,0:2}},{arrow:0,segment:{0:2,1:2}},{arrow:0,segment:{0:1,1:2}},{arrow:1,segment:{-1:1,0:1,-2:1}}],[32mblocks[0m:{[33m[-2,1,1,0][0m:[-26,-30,-28,-32,-27,-31,-29,-33],[33m[-1,0,1,0][0m:[-22,-24,-23,-25],[33m[-1,1,0,0][0m:[-20,-21],[33m[-1,1,1,-1][0m:[12,14,13,15],[33m[0,0,0,0][0m:[16,18,17,19],[33m[0,0,1,-1][0m:[-4,-8,-6,-10,-5,-9,-7,-11],[33m[0,1,0,-1][0m:[-0,-2,-1,-3],[33m[0,1,1,-2][0m:[-34,-38,-36,-40,-35,-39,-37,-41]}}
{[32mnames[0m:[D,C,B,A],[32medges[0m:[{arrow:1,segment:{-2:2,-1:1,0:2}},{arrow:0,segment:{0:2,1:2}},{arrow:0,segment:{0:1,1:2}},{arrow:1,segment:{-1:1,0:1,-2:1}}],[32mblocks[0m:{[33m[-2,1,1,0][0m:[-26,-30,-28,-32,-27,-31,-29,-33],[33m[-1,0,1,0][0m:[-22,-24,-23,-25],[33m[-1,1,0,0][0m:[-20,-21],[33m[-1,1,1,-1][0m:[12,14,13,15],[33m[0,0,0,0][0m:[16,18,17,19],[33m[0,0,1,-1][0m:[-4,-8,-6,-10,-5,-9,-7,-11],[33m[0,1,0,-1][0m:[0,-2,-1,-3],[33m[0,1,1,-2][0m:[-34,-38,-36,-40,-35,-39,-37,-41]}}
{[32mnames[0m:[D,C,B,A],[32medges[0m:[{arrow:1,segment:{-2:2,-1:1,0:2}},{arrow:0,segment:{0:2,1:2}},{arrow:0,segment:{0:1,1:2}},{arrow:1,segment:{-1:1,0:1,-2:1}}],[32mblocks[0m:{[33m[-2,1,1,0][0m:[-26,-30,-28,-32,-27,-31,-29,-33],[33m[-1,0,1,0][0m:[-22,-24,-23,-25],[33m[-1,1,0,0][0m:[-20,-21],[33m[-1,1,1,-1][0m:[12,14,13,15],[33m[0,0,0,0][0m:[16,18,17,19],[33m[0,0,1,-1][0m:[-4,-8,-6,-10,-5,-9,-7,-11],[33m[0,1,0,-1][0m:[0,-2,-1,-3],[33m[0,1,1,-2][0m:[-34,-38,-36,-40,-35,-39,-37,-41]}}
